File: APE-05 Creación, Lectura y Escritura de Ficheros/Ejercicio 8.cpp
#include <fstream>
#include <iostream>
#include <sstream>
#include <string>
#include <vector>
#include <limits>

const std::string ARCHIVO_LIBROS = "biblioteca.txt";

class Libro {
public:
    Libro(const std::string &codigo, const std::string &titulo, const std::string &autor, const std::string &estado)
        : codigo(codigo), titulo(titulo), autor(autor), estado(estado) {}

    const std::string &getCodigo() const { return codigo; }
    const std::string &getTitulo() const { return titulo; }
    const std::string &getAutor() const { return autor; }
    const std::string &getEstado() const { return estado; }

    bool estaDisponible() const {
        return estado == "disponible";
    }

    void prestar() {
        estado = "prestado";
    }

    void devolver() {
        estado = "disponible";
    }

    std::string toRegistro() const {
        std::ostringstream out;
        out << codigo << "|" << titulo << "|" << autor << "|" << estado;
        return out.str();
    }

    std::string toString() const {
        std::ostringstream out;
        out << "Código: " << codigo << "\n"
            << "Título: " << titulo << "\n"
            << "Autor: " << autor << "\n"
            << "Estado: " << estado;
        return out.str();
    }

private:
    std::string codigo;
    std::string titulo;
    std::string autor;
    std::string estado;
};

int leerEntero(const std::string &mensaje) {
    while (true) {
        std::cout << mensaje;
        int valor;
        if (std::cin >> valor) {
            std::cin.ignore(std::numeric_limits<std::streamsize>::max(), '\n');
            return valor;
        }
        std::cout << "Debe ingresar un número válido." << std::endl;
        std::cin.clear();
        std::cin.ignore(std::numeric_limits<std::streamsize>::max(), '\n');
    }
}

std::string leerTextoNoVacio(const std::string &mensaje) {
    while (true) {
        std::cout << mensaje;
        std::string texto;
        std::getline(std::cin, texto);
        if (!texto.empty()) {
            return texto;
        }
        std::cout << "El texto no puede estar vacío." << std::endl;
    }
}

std::vector<Libro> cargarLibros() {
    std::vector<Libro> libros;
    std::ifstream archivo(ARCHIVO_LIBROS);
    if (!archivo.is_open()) {
        return libros;
    }

    std::string linea;
    while (std::getline(archivo, linea)) {
        std::istringstream ss(linea);
        std::string codigo, titulo, autor, estado;
        if (std::getline(ss, codigo, '|') && std::getline(ss, titulo, '|') && std::getline(ss, autor, '|') && std::getline(ss, estado)) {
            libros.emplace_back(codigo, titulo, autor, estado);
        }
    }
    return libros;
}

void guardarLibros(const std::vector<Libro> &libros) {
    std::ofstream archivo(ARCHIVO_LIBROS);
    if (!archivo.is_open()) {
        std::cout << "Error al guardar libros." << std::endl;
        return;
    }
    for (const auto &libro : libros) {
        archivo << libro.toRegistro() << "\n";
    }
}

Libro *buscarLibro(std::vector<Libro> &libros, const std::string &codigo) {
    for (auto &libro : libros) {
        if (libro.getCodigo() == codigo) {
            return &libro;
        }
    }
    return nullptr;
}

void mostrarMenu() {
    std::cout << "=== Biblioteca Virtual ===" << std::endl;
    std::cout << "1. Agregar libro" << std::endl;
    std::cout << "2. Buscar libro" << std::endl;
    std::cout << "3. Mostrar libros disponibles" << std::endl;
    std::cout << "4. Mostrar libros prestados" << std::endl;
    std::cout << "5. Prestar libro" << std::endl;
    std::cout << "6. Devolver libro" << std::endl;
    std::cout << "7. Mostrar todos los libros" << std::endl;
    std::cout << "8. Salir" << std::endl;
}

void agregarLibro(std::vector<Libro> &libros) {
    std::string codigo = leerTextoNoVacio("Código del libro: ");
    if (buscarLibro(libros, codigo) != nullptr) {
        std::cout << "El código ya existe en la biblioteca." << std::endl;
        return;
    }
    std::string titulo = leerTextoNoVacio("Título del libro: ");
    std::string autor = leerTextoNoVacio("Autor del libro: ");
    libros.emplace_back(codigo, titulo, autor, "disponible");
    guardarLibros(libros);
    std::cout << "Libro agregado exitosamente." << std::endl;
}

void buscarLibro(const std::vector<Libro> &libros) {
    std::string codigo = leerTextoNoVacio("Ingrese el código del libro a buscar: ");
    for (const auto &libro : libros) {
        if (libro.getCodigo() == codigo) {
            std::cout << libro.toString() << std::endl;
            return;
        }
    }
    std::cout << "Libro no encontrado." << std::endl;
}

void mostrarDisponibles(const std::vector<Libro> &libros) {
    bool encontrado = false;
    std::cout << "=== Libros Disponibles ===" << std::endl;
    for (const auto &libro : libros) {
        if (libro.estaDisponible()) {
            std::cout << libro.toString() << std::endl << std::endl;
            encontrado = true;
        }
    }
    if (!encontrado) {
        std::cout << "No hay libros disponibles en este momento." << std::endl;
    }
}

void mostrarPrestados(const std::vector<Libro> &libros) {
    bool encontrado = false;
    std::cout << "=== Libros Prestados ===" << std::endl;
    for (const auto &libro : libros) {
        if (!libro.estaDisponible()) {
            std::cout << libro.toString() << std::endl << std::endl;
            encontrado = true;
        }
    }
    if (!encontrado) {
        std::cout << "No hay libros prestados actualmente." << std::endl;
    }
}

void prestarLibro(std::vector<Libro> &libros) {
    std::string codigo = leerTextoNoVacio("Ingrese el código del libro a prestar: ");
    Libro *libro = buscarLibro(libros, codigo);
    if (libro == nullptr) {
        std::cout << "Libro no encontrado." << std::endl;
    } else if (!libro->estaDisponible()) {
        std::cout << "El libro ya está prestado." << std::endl;
    } else {
        libro->prestar();
        guardarLibros(libros);
        std::cout << "Libro prestado exitosamente." << std::endl;
    }
}

void devolverLibro(std::vector<Libro> &libros) {
    std::string codigo = leerTextoNoVacio("Ingrese el código del libro a devolver: ");
    Libro *libro = buscarLibro(libros, codigo);
    if (libro == nullptr) {
        std::cout << "Libro no encontrado." << std::endl;
    } else if (libro->estaDisponible()) {
        std::cout << "El libro ya está disponible." << std::endl;
    } else {
        libro->devolver();
        guardarLibros(libros);
        std::cout << "Libro devuelto exitosamente." << std::endl;
    }
}

void mostrarTodos(const std::vector<Libro> &libros) {
    if (libros.empty()) {
        std::cout << "La biblioteca está vacía." << std::endl;
        return;
    }
    std::cout << "=== Todos los Libros ===" << std::endl;
    for (const auto &libro : libros) {
        std::cout << libro.toString() << std::endl << std::endl;
    }
}

int main() {
    std::vector<Libro> libros = cargarLibros();
    int opcion;
    do {
        mostrarMenu();
        opcion = leerEntero("Seleccione una opción: ");
        switch (opcion) {
            case 1:
                agregarLibro(libros);
                break;
            case 2:
                buscarLibro(libros);
                break;
            case 3:
                mostrarDisponibles(libros);
                break;
            case 4:
                mostrarPrestados(libros);
                break;
            case 5:
                prestarLibro(libros);
                break;
            case 6:
                devolverLibro(libros);
                break;
            case 7:
                mostrarTodos(libros);
                break;
            case 8:
                std::cout << "Gracias por usar la Biblioteca Virtual. ¡Hasta luego!" << std::endl;
                break;
            default:
                std::cout << "Opción inválida. Intente de nuevo." << std::endl;
        }
        std::cout << std::endl;
    } while (opcion != 8);
    return 0;
}

File: APE-05 Creación, Lectura y Escritura de Ficheros/Ejercicio8.java
import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.File;
import java.io.FileReader;
import java.io.FileWriter;
import java.io.IOException;
import java.util.ArrayList;
import java.util.List;
import java.util.Scanner;

public class Ejercicio8 {

    private static final String ARCHIVO_LIBROS = "biblioteca.txt";
    private static List<Libro> libros = new ArrayList<>();

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        cargarLibros();

        int opcion;
        do {
            mostrarMenu();
            opcion = leerEntero(scanner, "Seleccione una opción: ");
            switch (opcion) {
                case 1:
                    agregarLibro(scanner);
                    break;
                case 2:
                    buscarLibro(scanner);
                    break;
                case 3:
                    mostrarDisponibles();
                    break;
                case 4:
                    mostrarPrestados();
                    break;
                case 5:
                    prestarLibro(scanner);
                    break;
                case 6:
                    devolverLibro(scanner);
                    break;
                case 7:
                    mostrarTodos();
                    break;
                case 8:
                    System.out.println("Gracias por usar la Biblioteca Virtual. ¡Hasta luego!");
                    break;
                default:
                    System.out.println("Opción inválida. Intente de nuevo.");
            }
            System.out.println();
        } while (opcion != 8);

        guardarLibros();
        scanner.close();
    }

    private static void mostrarMenu() {
        System.out.println("=== Biblioteca Virtual ===");
        System.out.println("1. Agregar libro");
        System.out.println("2. Buscar libro");
        System.out.println("3. Mostrar libros disponibles");
        System.out.println("4. Mostrar libros prestados");
        System.out.println("5. Prestar libro");
        System.out.println("6. Devolver libro");
        System.out.println("7. Mostrar todos los libros");
        System.out.println("8. Salir");
    }

    private static int leerEntero(Scanner scanner, String mensaje) {
        while (true) {
            System.out.print(mensaje);
            String linea = scanner.nextLine().trim();
            try {
                return Integer.parseInt(linea);
            } catch (NumberFormatException e) {
                System.out.println("Debe ingresar un número válido.");
            }
        }
    }

    private static String leerTextoNoVacio(Scanner scanner, String mensaje) {
        while (true) {
            System.out.print(mensaje);
            String texto = scanner.nextLine().trim();
            if (!texto.isEmpty()) {
                return texto;
            }
            System.out.println("El texto no puede estar vacío.");
        }
    }

    private static void agregarLibro(Scanner scanner) {
        String codigo = leerTextoNoVacio(scanner, "Código del libro: ");
        if (buscarPorCodigo(codigo) != null) {
            System.out.println("El código ya existe en la biblioteca.");
            return;
        }
        String titulo = leerTextoNoVacio(scanner, "Título del libro: ");
        String autor = leerTextoNoVacio(scanner, "Autor del libro: ");
        Libro libro = new Libro(codigo, titulo, autor, "disponible");
        libros.add(libro);
        guardarLibros();
        System.out.println("Libro agregado exitosamente.");
    }

    private static void buscarLibro(Scanner scanner) {
        String codigo = leerTextoNoVacio(scanner, "Ingrese el código del libro a buscar: ");
        Libro libro = buscarPorCodigo(codigo);
        if (libro != null) {
            System.out.println(libro);
        } else {
            System.out.println("Libro no encontrado.");
        }
    }

    private static Libro buscarPorCodigo(String codigo) {
        for (Libro libro : libros) {
            if (libro.getCodigo().equalsIgnoreCase(codigo)) {
                return libro;
            }
        }
        return null;
    }

    private static void mostrarDisponibles() {
        List<Libro> disponibles = new ArrayList<>();
        for (Libro libro : libros) {
            if (libro.isDisponible()) {
                disponibles.add(libro);
            }
        }

        if (disponibles.isEmpty()) {
            System.out.println("No hay libros disponibles en este momento.");
        } else {
            System.out.println("=== Libros Disponibles ===");
            for (int i = 0; i < disponibles.size(); i++) {
                System.out.printf("%d. %s%n", i + 1, disponibles.get(i));
            }
        }
    }

    private static void mostrarPrestados() {
        List<Libro> prestados = new ArrayList<>();
        for (Libro libro : libros) {
            if (!libro.isDisponible()) {
                prestados.add(libro);
            }
        }

        if (prestados.isEmpty()) {
            System.out.println("No hay libros prestados actualmente.");
        } else {
            System.out.println("=== Libros Prestados ===");
            for (int i = 0; i < prestados.size(); i++) {
                System.out.printf("%d. %s%n", i + 1, prestados.get(i));
            }
        }
    }

    private static void mostrarTodos() {
        if (libros.isEmpty()) {
            System.out.println("La biblioteca está vacía.");
        } else {
            System.out.println("=== Todos los Libros ===");
            for (int i = 0; i < libros.size(); i++) {
                System.out.printf("%d. %s%n", i + 1, libros.get(i));
            }
        }
    }

    private static void prestarLibro(Scanner scanner) {
        String codigo = leerTextoNoVacio(scanner, "Ingrese el código del libro a prestar: ");
        Libro libro = buscarPorCodigo(codigo);
        if (libro == null) {
            System.out.println("Libro no encontrado.");
        } else if (!libro.isDisponible()) {
            System.out.println("El libro ya está prestado.");
        } else {
            libro.prestar();
            guardarLibros();
            System.out.println("Libro prestado exitosamente.");
        }
    }

    private static void devolverLibro(Scanner scanner) {
        String codigo = leerTextoNoVacio(scanner, "Ingrese el código del libro a devolver: ");
        Libro libro = buscarPorCodigo(codigo);
        if (libro == null) {
            System.out.println("Libro no encontrado.");
        } else if (libro.isDisponible()) {
            System.out.println("El libro ya está disponible.");
        } else {
            libro.devolver();
            guardarLibros();
            System.out.println("Libro devuelto exitosamente.");
        }
    }

    private static void cargarLibros() {
        File archivo = new File(ARCHIVO_LIBROS);
        if (!archivo.exists()) {
            return;
        }

        try (BufferedReader reader = new BufferedReader(new FileReader(archivo))) {
            String linea;
            while ((linea = reader.readLine()) != null) {
                String[] partes = linea.split("\\|");
                if (partes.length == 4) {
                    String codigo = partes[0];
                    String titulo = partes[1];
                    String autor = partes[2];
                    String estado = partes[3];
                    libros.add(new Libro(codigo, titulo, autor, estado));
                }
            }
        } catch (IOException e) {
            System.out.println("Error al cargar libros: " + e.getMessage());
        }
    }

    private static void guardarLibros() {
        try (BufferedWriter writer = new BufferedWriter(new FileWriter(ARCHIVO_LIBROS))) {
            for (Libro libro : libros) {
                writer.write(libro.toRegistro());
                writer.newLine();
            }
        } catch (IOException e) {
            System.out.println("Error al guardar libros: " + e.getMessage());
        }
    }
}

class Libro {

    private final String codigo;
    private final String titulo;
    private final String autor;
    private String estado;

    public Libro(String codigo, String titulo, String autor, String estado) {
        this.codigo = codigo;
        this.titulo = titulo;
        this.autor = autor;
        this.estado = estado;
    }

    public String getCodigo() {
        return codigo;
    }

    public boolean isDisponible() {
        return estado.equalsIgnoreCase("disponible");
    }

    public void prestar() {
        estado = "prestado";
    }

    public void devolver() {
        estado = "disponible";
    }

    public String toRegistro() {
        return codigo + "|" + titulo + "|" + autor + "|" + estado;
    }

    @Override
    public String toString() {
        return String.format("[%s] %s - %s (%s)", codigo, titulo, autor, estado);
    }
}
